File: apps/mobile/lib/features/circles/domain/circle.dart
import 'circle_member.dart';

class Circle {
  const Circle({
    required this.id,
    required this.name,
    required this.ownerId,
    required this.createdAt,
    required this.deletedAt,
  });

  factory Circle.fromJson(Map<String, dynamic> json) {
    return Circle(
      id: json['id'] as String,
      name: json['name'] as String,
      ownerId: json['ownerId'] as String,
      createdAt: json['createdAt'] as String,
      deletedAt: json['deletedAt'] as String?,
    );
  }

  final String id;
  final String name;
  final String ownerId;
  final String createdAt;
  final String? deletedAt;
}

class CircleWithMembers extends Circle {
  const CircleWithMembers({
    required super.id,
    required super.name,
    required super.ownerId,
    required super.createdAt,
    required super.deletedAt,
    required this.members,
  });

  factory CircleWithMembers.fromJson(Map<String, dynamic> json) {
    final membersJson = json['members'] as List<dynamic>;
    return CircleWithMembers(
      id: json['id'] as String,
      name: json['name'] as String,
      ownerId: json['ownerId'] as String,
      createdAt: json['createdAt'] as String,
      deletedAt: json['deletedAt'] as String?,
      members: membersJson
          .map((m) => CircleMember.fromJson(m as Map<String, dynamic>))
          .toList(),
    );
  }

  final List<CircleMember> members;
}

File: packages/shared-types/src/circle.ts
export interface Circle {
  id: string;
  name: string;
  ownerId: string;
  /**
   * The code somebody types to join this circle. Only ever populated for
   * the circle's owner — every other member receives `null`, matching the
   * server-side rule that only the owner can bring people in.
   */
  inviteCode: string | null;
  createdAt: string;
  deletedAt: string | null;
}

export interface CircleMember {
  circleId: string;
  userId: string;
  username: string;
  role: 'owner' | 'member';
  joinedAt: string;
}

export interface CircleWithMembers extends Circle {
  members: CircleMember[];
}

export interface CreateCircleRequest {
  name: string;
}

export interface UpdateCircleRequest {
  name: string;
}

export interface AddCircleMemberRequest {
  username: string;
}

export interface JoinCircleRequest {
  code: string;
}

export interface AdminCircleSummary {
  id: string;
  name: string;
  ownerId: string;
  memberCount: number;
  createdAt: string;
}

export interface Geofence {
  id: string;
  circleId: string;
  name: string;
  center: GeoPoint;
  radiusMeters: number;
  createdBy: string;
}

export interface GeoPoint {
  lat: number;
  lng: number;
}
